src/libraries/Microsoft.PowerFx.Core/Public/DisplayNames/RenameDriver.cs

```diff
@@ -0,0 +1,70 @@
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.Glue;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core
+{
+    public class RenameDriver
+    {
+        private readonly RecordType _parameters;
+        private readonly SimpleResolver _resolver;
+
+        internal RenameDriver(RecordType parameters, DPath pathToRename, DName updatedName, SimpleResolver resolver)
+        {
+            var segments = new Queue<DName>(pathToRename.Segments());
+            Contracts.CheckParam(segments.Count > 0, nameof(parameters));
+
+            // After this point, _parameters should have at most one logical->display pair that can change in this conversion
+            _parameters = RenameFormulaTypeHelper(parameters, segments, updatedName) as RecordType;
+            _resolver = resolver;
+        }
+
+        /// <summary>
+        /// Applies rename operation to <paramref name="expressionText"/>.
+        /// </summary>
+        /// <param name="expressionText">Expression in which to rename the parameter field.</param>
+        /// <returns>Expression with rename applied.</returns>
+        public string ApplyRename(string expressionText)
+        {
+            return Engine.ConvertExpression(expressionText, _parameters, _resolver, true);
+        }
+
+        private static FormulaType RenameFormulaTypeHelper(AggregateType nestedType, Queue<DName> segments, DName updatedName)
+        {
+            var field = segments.Dequeue();
+            if (segments.Count == 0)
+            {
+                // Create a display name provider with only the name in question
+                var names = new Dictionary<DName, DName>
+                {
+                    [field] = updatedName
+                };
+                var newProvider = new SingleSourceDisplayNameProvider(names);
+
+                return FormulaType.Build(DType.ReplaceDisplayNameProvider(DType.DisableDisplayNameProviders(nestedType._type), newProvider));
+            }
+
+            var fieldType = nestedType.MaybeGetFieldType(field);
+            if (fieldType is not AggregateType aggregateType)
+            {
+                // Path doesn't exist within parameters, return as is, stripping displaynameproviders
+                return FormulaType.Build(DType.DisableDisplayNameProviders(nestedType._type));
+            }
+
+            var updatedType = RenameFormulaTypeHelper(aggregateType, segments, updatedName);
+            var fError = false;
+
+            // Use DType internals to swap one field type for the updated one and disable all other display names
+            var dropped = DType.DisableDisplayNameProviders(nestedType._type.Drop(ref fError, DPath.Root, field));
+            Contracts.Assert(!fError);
+
+            return FormulaType.Build(dropped.Add(field, updatedType._type));
```

Comment:
At this point the queue should be empty right?

Reply:
yes, after the recursive call, the queue is empty.

Reply:
or we hit the case where the path doesn't fully exist in the type, in which case the queue may still have values in it but we  stopped recursing early.